external-crates/move/tools/move-package/src/resolution/dependency_graph.rs

```diff
@@ -253,13 +217,49 @@ impl<Progress: Write> DependencyGraphBuilder<Progress> {
             DependencyMode::Always,
             &root_manifest.dependencies,
         )?;
+        let dep_lock_files = dep_graphs
+            .values()
+            .map(|graph_info| graph_info.g.write_to_lock(self.install_dir.clone()))
+            .collect::<Result<Vec<LockFile>>>()?;
         let dev_dep_graphs = self.collect_graphs(
             parent,
             root_manifest.package.name,
             root_path.clone(),
             DependencyMode::DevOnly,
             &root_manifest.dev_dependencies,
         )?;
+        let dev_dep_lock_files = dev_dep_graphs
+            .values()
+            .map(|graph_info| graph_info.g.write_to_lock(self.install_dir.clone()))
+            .collect::<Result<Vec<LockFile>>>()?;
+
+        let new_deps_digest = self.dependency_digest(dep_lock_files, dev_dep_lock_files)?;
+
+        let (manifest_digest, deps_digest) = match old_manifest_digest_opt {
+            None => (new_manifest_digest, new_deps_digest),
+            Some(old_manifest_digest) => {
+                if old_manifest_digest == new_manifest_digest {
+                    // manifest digests match - see if dependency digests match as well
+                    if let Some(old_deps_digest) = old_deps_digest_opt {
+                        // old dependencies digest exist - see if it matches
+                        if old_deps_digest == new_deps_digest {
+                            // both manifest digest and dependency digest match - we can simply read
+                            // the old lock file
+                            return Ok((
+                                DependencyGraph::read_from_lock(
+                                    root_path,
+                                    root_manifest.package.name,
+                                    &mut lock_string.unwrap().as_bytes(), // safe since old_deps_digest exists
+                                    None,
+                                )?,
+                                false,
+                            ));
+                        } // else dependency digests do not match - re-assemble the graph
+                    } // else dependency digest in the lock file does not exist - re-assemble the graph
+                } // else manifest digests do not match - re-assemble the graph
+                (new_manifest_digest, new_deps_digest)
+            }
+        };
```

Comment:
nit: we can flatten this out a bit.  In fact, with this change, instead of returning three separate `Option`s when fetching the old digests and lock file content, you can return a single `Some((manifest_digest, deps_digest, lock_string))`to simplify things further:

```suggestion
        let dev_dep_lock_files = dev_dep_graphs
            .values()
            .map(|graph_info| graph_info.g.write_to_lock(self.install_dir.clone()))
            .collect::<Result<Vec<LockFile>>>()?;

        let new_deps_digest = self.dependency_digest(dep_lock_files, dev_dep_lock_files)?;

        let (manifest_digest, deps_digest) = match (old_manifest_digest_opt, old_deps_digest_opt, lock_string) {
            (Some(old_manifest_digest), Some(old_deps_digest), Some(lock_string)) 
                if old_manifest_digest == new_manifest_digest &&
                    old_deps_digest == new_deps_digest => {
                return Ok((
                    DependencyGraph::read_from_lock(
                        root_path,
                        root_manifest.package.name,
                        &mut lock_string.as_bytes(), // safe since old_deps_digest exists
                        None,
                    )?,
                    false,
                );
            }
            
            _ => (new_manifest_digest, new_deps_digest),
        };
```